File: docs/scripting/samples/python-scripts/python-examples/Event_Run_Order_Parameters.py
"""
Event Run Order Parameters

Python scripting example demonstrating Cellario API integration and automation capabilities
"""

#Cellario modules
from HRB.Cellario.Scripting import *
from HRB.Cellario.Scripting.API import *


# Method called during Cellario protocol execution when a sample arrives at the scripting step.
# <remarks>Executes synchronously with the run scheduler. Device operation results not available
# until <see cref="ReleaseResources"/> method is called.</remarks>
# <Object name="api">Access to properties and methods for interacting with the Cellario run-time scheduler, 
# samples, resources and devices operations.</param>
def Execute(api : PythonScriptingApi):
    # run time parameter name
    reader_output_directory_name = "Output File Directory";
    
    # creation of a message about order run information
    run_order_id = api.CurrentPlate.CurrentRun.RunNumber
    run_order_desc = api.CurrentPlate.CurrentRun.Description
    protocol_name = api.CurrentPlate.CurrentProtocol.ProtocolName
    
    message = ("Before Order State Change for run = " + str(run_order_id) + " \n"
              + "Order Description '" + str(run_order_desc) + "' \n"
              + "Protocol Name - " + protocol_name + " \n")
    api.Messaging.Notify("Before Order State Change", message)
    
    # number of parameters in order run
    parameters = api.CurrentPlate.CurrentRun.RunOrderParameters
    # TBD api.Messaging.Notify("Run Order Parameter Count", "Found - ", str(len(parameters)))
    
    index = 0
    new_param_value = None
    for e in parameters:
        index += 1
        pname = e.ParameterName
        ppid = e.ProtocolParameterId
        pvalue = e.ParameterValue
        roparamid = e.RunOrderParameterId
        
        message = ("Parameter  " + str(index) + " Name is '" + str(pname) + "' \n"
              + "Protocol Parameter Id is '" + str(ppid) + "' \n"
              + "Parameter Value is '" + str(pvalue) + " \n"
              + "Run Order Parameter Id is '" + str(roparamid) + " \n")
        api.Messaging.Notify("Run Order Parameter ", message)
        
        if (pname == reader_output_directory_name):
            prot_param_list = api.CurrentPlate.CurrentProtocol.Parameters
            prot_param = next(c for c in prot_param_list if c.ProtocolParameterId == ppid)
            
            choices_list = prot_param.ChoicesList
            new_param_value = choices_list[len(choices_list) - 1]
            e.ParameterValue = new_param_value
            
            message = ("Parameter  " + str(index) + " Name '" + str(pname) + "' \n"
               + "Changed value from '" + str(pvalue) + "' to '" + str(new_param_value) + "' \n")
              
            api.Messaging.Notify("Run Order Parameter ", message)

  

File: docs/scripting/samples/csharp-scripts/examples/Event_Run_Order_Parameters.cs
/**
 * Event Run Order Parameters
 * 
 * Example script showing basic Cellario protocol automation and customization features
 */

#region Header
// ** Copyright HighRes Biosolutions, Inc. 2020 **
// 
//    File:      EventScript_RunOrderParameters.cs 
//    Project:   ScriptLibrary
//    Solution:  CellarioWPF
#endregion

using System;
using System.Linq;
using HRB.Cellario.Scripting.API;

namespace HRB.Cellario.Scripting.ScriptLibrary
{
    public class EventScript_RunOrderParameters : AbstractScript
    {

        /// <summary>
        /// Method called during Cellario protocol execution when a sample arrives at the scripting step.
        /// </summary>
        /// <remarks>Executes synchronously with the run scheduler. Device operation results not available
        /// until <see cref="ReleaseResources"/> method is called.</remarks>
        /// <param name="api">Access to properties and methods for interacting with the Cellario run-time scheduler, 
        /// samples, resources and devices operations.</param>
        public override void Execute(IScriptingApi api)
        {

            var readerOutputDirectoryName = "Output File Directory";
            //about this script 
            var runorderId = api.CurrentRun.RunNumber;
            var runOrderDesc = api.CurrentRun.Description;
            var protocolName = api.CurrentRun.Protocol.ProtocolName;

            var message = "Before Order State Change for run = " + runorderId.ToString() + " \n";
            message += "Order Description '" + runOrderDesc + "' \n";
            message += "Order State - " + api.CurrentRun.RunState + " \n";
            message += "Protocol Name - " + protocolName + " \n";
            api.Messaging.Notify("Before Order State Change", message);

            var parameters = api.CurrentRun.RunOrderParameters.ToList();
            api.Messaging.Notify("Run Order Parameter Count", "Found - " + parameters.Count().ToString());

            var index = 0;
            var newParamValue = string.Empty;
            foreach (var parameter in parameters)
            {
                index++;
                var pname = parameter.ParameterName;
                var ppid = parameter.ProtocolParameterId;
                var pvalue = parameter.ParameterValue;
                var roparamid = parameter.RunOrderParameterId;

                message = "Parameter " + index.ToString() + " Name is '" + pname + "' \n";
                message += "Protocol Parameter Id is '" + ppid + "' \n";
                message += "Parameter Value is '" + pvalue + "' \n";
                message += "Run Order Parameter Id is '" + roparamid + "' \n";

                api.Messaging.Notify("Run Order Parameter " + index.ToString(), message);

                if (pname == readerOutputDirectoryName)
                {
                    var protParam = api.CurrentRun.Protocol.Parameters.SingleOrDefault(a => a.ProtocolParameterId == ppid);
                    var fppChoicesList = protParam.ChoicesList;
                    newParamValue = fppChoicesList.Last();
                    parameter.ParameterValue = newParamValue;
                    message = "Parameter " + index.ToString() + " Name '" + pname + "' \n";
                    message += "Changed value from '" + pvalue + "' to '" + newParamValue + "' \n";

                    api.Messaging.Notify("Run Order Parameter " + index.ToString(), message);
                }
            }
        }
    }
}
